src/shared/network-cloud-util.c

```diff
@@ -0,0 +1,280 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "in-addr-util.h"
+#include "ether-addr-util.h"
+#include "network-cloud-util.h"
+#include "parse-util.h"
+
+int network_cloud_manager_new(CloudProvider provider, const char *url, NetworkCloudManager **ret) {
+        _cleanup_(network_cloud_manager_freep) NetworkCloudManager *m = NULL;
+
+        m = new(NetworkCloudManager, 1);
+        if (!m)
+                return -ENOMEM;
+
+        *m = (NetworkCloudManager) {
+                .provider = provider,
+                .curl = curl_easy_init(),
+        };
+
+        if (!m->curl)
+                return -ENOSR;
+
+        if (url)
+                curl_easy_setopt(m->curl, CURLOPT_URL, url);
+
+        m->interfaces_by_mac = ordered_hashmap_new(&ether_addr_hash_ops);
+        if (!m->interfaces_by_mac)
+                return -ENOMEM;
+
+        *ret = TAKE_PTR(m);
+        return 0;
+}
+
+void *network_cloud_manager_free(NetworkCloudManager *m) {
+        NetworkCloudMetaData *i;
+
+        if (!m)
+                return NULL;
+
+        if (m->curl)
+                curl_easy_cleanup(m->curl);
+
+        if (m->request_header)
+                curl_slist_free_all(m->request_header);
+
+
+        ORDERED_HASHMAP_FOREACH(i, m->interfaces_by_mac) {
+                ordered_hashmap_free_free(i->ipv4);
+                network_cloud_metadata_free(i);
+        }
+
+        ordered_hashmap_free(m->interfaces_by_mac);
+
+        return mfree(m);
+}
+
+int network_cloud_metadata_new(NetworkCloudMetaData **ret) {
+        _cleanup_(network_cloud_metadata_freep) NetworkCloudMetaData *m = NULL;
+
+        m = new0(NetworkCloudMetaData, 1);
+        if (!m)
+                return -ENOMEM;
+
+        m->ipv4 = ordered_hashmap_new(&trivial_hash_ops);
+        if (!m->ipv4)
+                return -ENOMEM;
```

Comment:
Just use `ordered_hashmap_ensure_put` when adding items to this hashmap, drop initialization here. Then the whole function is equivalent to `new0` and can be dropped.